### PR TITLE
feat: refactors to convert nuxt into a monorepo

### PR DESCRIPTION
**Pull request by clarkdo**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Changes summary: https://github.com/nuxt/nuxt.js/pull/4051#issuecomment-429611188

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by galvez** [comment]
@clarkdo looking good!

**Comment by pi0** [comment]
Seems we have lots' of conflicts now :(

**Comment by clarkdo** [comment]
@pi0 Could you help me sync this branch with dev branch? The confilcts should be solved automatically

**Review by galvez — APPROVED** [review]

**Review by TheAlexLichter — APPROVED** [review]

**Comment by pi0** [comment]
Changes summary: 

- Using Lerna and Yarn workspaces
- Prefer named exports for all packages
- Expose both `main` and `module` fields for each package (Fully ESM)
- Created specific packages for nuxt _distributions_:
  - `nuxt`
  - `nuxt-start`
  - `nuxt-legacy`
- Split nuxt into sub-packages:
  - `@nuxtjs/app`
  - `@nuxtjs/builder`
  - `@nuxtjs/core`
  - `@nuxtjs/common`
- Add `options.build.template` (replaces `options.nuxtAppDir`) so we can finally have a way to implement non-web/default templates (Weex, NativeScript, Framework7, etc) (Maybe react one day :D)

**Comment by pi0** [comment]
TODO:

- [ ] We need a build step for development again! Either implement watch mode for `scripts/package.js` or Make `scripts/nuxt` working without dist files! (ESM bug?)
- [ ] Tests failing
- [ ] Review by all of core team members

**Comment by pi0** [comment]
Some changes: module field and `src` directory will not be in published npm packages. Mainly because of the lack of proper ES support ATM. It also reduced packages size and complexity to handle.

Also currently we don't have much coverage for `@nuxtjs/cli` package (Like before) because Jest can't instrument `spawn`ed processes. I think having coverage can be achieved easily because of new CLI low-level command exports (See `packages/cli/src/index.js` and `packages/cli/commands`)

File: scripts/package.js [hunk 1]
```diff
@@ -120,34 +120,68 @@ export default class Package extends EventEmitter {
...
-    }, this.pkg))
+    }, this.pkg)
+
+    if (_watch) {
+      // Watch
+      const watcher = watch(config)
+      watcher.on('event', (event) => {
+        switch (event.code) {
+          // The watcher is (re)starting
+          case 'START': return this.logger.debug('Watching for changes')
```
Comment (manniL) [inline scripts/package.js]: Could we "reference" the event codes somehow? It's unlikely that they'll break but it'd be *safer* to have a Map or a lookup object 🤔

File: packages/builder/src/builder.js [hunk 2]
```diff
@@ -1,22 +1,23 @@
 import path from 'path'
 import fs from 'fs'
-
 import pify from 'pify'
```
Comment (manniL) [inline packages/builder/src/builder.js]: @pi0 Can we replace it with `utils.promisify` or do we need it for the legacy build? 🤔

Comment (pi0) [reply packages/builder/src/builder.js]: I'm on it ;)

Comment (pi0) [reply packages/builder/src/builder.js]: @manniL Just created and published [pify-native](https://github.com/blindmedia/pify-native). But it seems that `pify` is way faster. So let's keep using it for now :) (https://github.com/sindresorhus/pify/issues/41#issuecomment-429988506)

Comment (manniL) [reply packages/builder/src/builder.js]: Alright, thanks for the info @pi0!

**Comment by galvez** [comment]
> because Jest can't instrument spawned processes

@pi0 I was eventually able to get it to work, that is, capture the output of buffered executables. I'll try and get it wired up after this PR gets in. Reading through.

**Comment by pi0** [comment]
Locally merged with 9c1e0d174319f209bbfa4d34cfd2e17e82f45537

**Pull request closed without merging**